### PR TITLE
file_readv_compat/file_writev_compat: Fix partial success handling

### PR DESCRIPTION
**Pull request by yamt**
## Summary

The problem has been inherited from the original libc readv/writev implementation. However, now it's exposed in more situations because this implemenation is used to back read/write as well.

I expect this fixes the regressions observed on the Espressif CI. https://github.com/apache/nuttx/pull/13498#issuecomment-2448031197

Note that, even with this fix, these "compat" readv/writev implementations are still inheritedly broken. (E.g. consider that a data boundary happens to match one of iovec boundaries) However, this fix is enough for read/write, where iovcnt is always 1.

## Impact

read/write on character devices or such.

## Testing

Tested on esp32-devkitc, with https://github.com/apache/nuttx/pull/14479 and other changes.

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR does not fully meet the NuttX requirements.  While it provides a summary of the problem and the solution, it lacks crucial details required by the template.

Here's a breakdown of missing information:

* **Summary:**  Missing a clear statement of *what* functional part of the code is being changed (e.g., `fs/readv.c`, libc implementation, etc.). The explanation of *how* the change works is also vague. "This fix is enough for read/write, where iovcnt is always 1" doesn't explain the actual code changes.  Crucially, the related NuttX issue number is missing.

* **Impact:** The impact section is severely incomplete.  It needs to address *all* the points in the template with specific YES/NO answers and descriptions where applicable.  Simply stating "read/write on character devices or such" doesn't adequately cover the potential impact.  Consider impacts on backward compatibility, documentation needs, and whether this is a new feature, a bug fix, or a performance improvement.

* **Testing:**  The testing information is insufficient. While a target is mentioned (esp32-devkitc),  the build host details are absent.  More importantly, there are no "Testing logs before change" and "Testing logs after change" as requested by the template.  Even a simple demonstration of the fix working would be beneficial.  Mentioning another PR adds complexity and makes it harder to assess the impact of *this* specific change in isolation.

**To make this PR compliant, the author needs to:**

1. **Clarify the Summary:** Specify the affected files and provide a more technical explanation of the code changes.  Include the related NuttX issue number.

2. **Complete the Impact Assessment:** Answer all the questions in the Impact section with YES/NO and provide details where necessary.

3. **Provide Adequate Testing Information:**  List the build host details (OS, CPU, compiler).  Include "before" and "after" logs demonstrating the fix, even if they are simple. Ideally, the test should be self-contained and not rely on other PRs.

**Comment by acassis** [comment]
@yamt do you have some suggestion to implement a fix that works for all cases?

@anchao @xiaoxiang781216 @pkarashchenko what do you think?

**Comment by yamt** [comment]
> @yamt do you have some suggestion to implement a fix that works for all cases?

please read the commit message of https://github.com/apache/nuttx/pull/13498

File: fs/vfs/fs_write.c [hunk 1]
```diff
@@ -68,35 +68,36 @@ static ssize_t file_writev_compat(FAR struct file *filep,
...
+      if (nwritten < 0)
+        {
+          return ntotal ? ntotal : nwritten;
+        }
 
-              if (nwritten < 0)
-                {
-                  return ntotal ? ntotal : nwritten;
-                }
+      ntotal    += nwritten;
```
Comment (xiaoxiang781216) [inline fs/vfs/fs_write.c:88]:
```suggestion
      ntotal += nwritten;
```

File: fs/vfs/fs_read.c [hunk 2]
```diff
@@ -68,42 +68,36 @@ static ssize_t file_readv_compat(FAR struct file *filep,
...
+      if (nread < 0)
+        {
+          return ntotal ? ntotal : nread;
+        }
 
-              else if (nread == 0)
-                {
-                  return ntotal;
-                }
+      ntotal    += nread;
```
Comment (xiaoxiang781216) [inline fs/vfs/fs_read.c:88]:
```suggestion
      ntotal += nread;
```

File: fs/vfs/fs_write.c (hunk 1 above)
Comment (yamt) [reply fs/vfs/fs_write.c]: done

File: fs/vfs/fs_read.c (hunk 2 above)
Comment (yamt) [reply fs/vfs/fs_read.c]: done

**Pull request merged**